hoot-core/src/main/cpp/hoot/core/ops/UnconnectedWaySnapper.cpp

```diff
@@ -579,6 +588,124 @@ bool UnconnectedWaySnapper::_snapUnconnectedWayEndNode(const NodePtr& unconnecte
   return snapOccurred;
 }
 
+void UnconnectedWaySnapper::_snapUnconnectedWayCrossings(const WayPtr& wayToSnap)
+{  // Only snap 1 to 2, no need to snap 2 to 1
+  if (wayToSnap->getStatus() != Status::Unknown1)
+    return;
+
+  // Get envelop of our way
+  std::shared_ptr<geos::geom::Envelope> env1(wayToSnap->getEnvelope(_map));
+
+  // Now find ways that are close to it
+  // FUTURE: if we need this to be faster, we could probably use a spatial index
+  WayMap ways = _map->getWays();
```

Comment:
The function `OsmMap::getWays()` returns a reference to the ways, if you capture it with a constant reference you don't have to create a copy of the way map which can be very expensive (both memory and speed-wize) when running on large datasets.  Some operations that modify the `WayMap` won't allow you to use a constant reference because it changes in the background, if this is the case, you can modify your loop to bring the actual snap out of the middle of the loop and keep a vector of IDs that pass the tests and then iterate that vector doing the snapping.  You can make the following change and if it crashes you know that your operation modifies the `WayMap` and can convert to the other way of doing it, extracting the work out of the search loop.
```suggestion
  const WayMap& ways = _map->getWays();
```